fastapi/src/ksi_oidc_fastapi/auth_router.py

```diff
@@ -0,0 +1,169 @@
+from .oidc_client import get_oidc_client
+
+import logging
+
+router = APIRouter(tags=["auth"])
+
+@router.get("/login")
+async def login(request: Request):
+    """Initiate OIDC login flow"""
+    session_key = get_or_create_session(request)
+    
+    
+    state = secrets.token_urlsafe(32)
+    nonce = secrets.token_urlsafe(32)
+    
+
+    session_manager.set_oauth_state(session_key, state, nonce)
+    
+
+    auth_url = get_oidc_client().get_authentication_url(
+        nonce=nonce,
+        state=state,
+        prompt_none=False
+    )
+    
+    return RedirectResponse(url=auth_url, status_code=302)
+
+@router.get("/callback")
+async def auth_callback(request: Request):
+    """Handle OIDC callback"""
+    session_key = get_or_create_session(request)
+    query_params = dict(request.query_params)
+    # logging.error(f"Query params: {query_params}")
+
+    if not session_manager.verify_oauth_state(session_key, query_params.get("state")):
+        logging.error("Invalid state parameter")
+        return RedirectResponse(
+            url=f"{router.prefix}/login",
+            status_code=302
+        )
+
+    session_data = session_manager.get_session(session_key)
+    if not session_data or not session_data.nonce:
+        logging.error("Invalid session state: Missing session data or nonce")
+        return JSONResponse(
+            content={
+                "error": "invalid_session",
+                "error_description": "Invalid session state. Please ensure cookies are enabled."
+            },
+            status_code=400
+        )
+
+    try:
+        auth_response = get_oidc_client().parse_authorization_callback_response(query_params)
+
+        tokens = get_oidc_client().exchange_code_for_access_token(
+            code=auth_response["code"],
+            expected_nonce=session_data.nonce
+        )
+
+        session_manager.set_user_authenticated(session_key, tokens)
+        return RedirectResponse(url="/", status_code=302)
+
+
+    except OidcProviderError as error:
+        logging.error(
+            f"Received OIDC provider error: {error.response.get('error')}, "
+            f"Description: {error.response.get('error_description')}"
+        )
+        
+        return JSONResponse(
+            content={
+                "error": "oidc_provider_error",
+                "error_description": error.response.get("error_description") or "Authentication failed",
+                "provider_error": error.response.get("error")
+            },
+            status_code=400
+        )
+
+    except OidcError as error:
+        logging.error(f"Got OIDC error in callback: {str(error)}", exc_info=True)
+        return JSONResponse(
+            content={
+                "error": "oidc_error",
+                "error_description": "An unexpected error occurred during authentication"
+            },
+            status_code=500
+        )
```

Comment:
Consider rendering the `callback_error.html` template instead of returning a JSON error response